Select-into identity issue fixed for bigint, numeric, decimal

deepakshi-mittal opened the pull request:
Added different functions to handle int, smallint and bigint input 
Added test cases covering join, all numeral-data types and temp tables 
Changed error message for unsupported datatypes(i.e. non-numeral types) 
Numeric and decimal to be treated as bigint, same as that of create table behavior with identity column

Task: BABEL-4315

### Issues Resolved
Select-into identity issue fixed for bigint, numeric, decimal

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

deepakshi-mittal committed:
Select-into identity issue fixed for bigint, numeric, decimal

Added different functions to handle int, smallint and bigint input
Added test cases covering join, all numeral-data types and temp tables
Changed error message for unsupported datatypes(ie non-numeral types)
Numeric and decimal to be treated as bigint, same as that of create
table behavior with identity column

Task: BABEL-4315

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

forestkeeper commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -236,29 +236,32 @@ TsqlFunctionIdentityInto(TypeName *typename, Node *seed, Node *increment, int lo
+		default:
+			ereport(ERROR,
 					(errcode(ERRCODE_SYNTAX_ERROR),
-					 errmsg("identity column type must be smallint, integer, bigint, or numeric")));
```
Did you add a test case for this error ?

deepakshi-mittal replied:
Yes, I added this one:

SELECT col1, IDENTITY(char, 1,1) AS id_num INTO babel_539NewTable1 FROM babel_539OldTable;
GO
~~ERROR (Code: 33557097)~~
~~ERROR (Message: identity column type must be of data type int, bigint, smallint, or decimal or numeric)~~

forestkeeper commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -5094,72 +5130,65 @@ static List *transformSelectIntoStmt(CreateTableAsStmt *stmt, const char *queryS
 					arg_num++;
 					switch (arg_num)
 					{
-						case 1:
```
We should throw some error when arg_num reach 4 or more in here ( although it's not possible , but for safe purpose)

forestkeeper commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -5064,12 +5065,47 @@ static char *get_oid_type_string(int type_oid){
+		case T_OpExpr:
+			opxpr = (OpExpr *)node;
+			if ((opxpr->args)->length != 1)
+				elog(ERROR, "Incorrect syntax near 'identity'");
```
We should have a test regarding this code block

deepakshi-mittal committed:
Added default case to throw error if number of function arguments are >=4
Added test case for expressions (1+1), (1-2)
Changed error type to syntax error

Task: BABEL-4315

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

forestkeeper reviewed (dismissed):
LGTM, make sure the non-error behavior is the same with Sql server, rest are OK

deepakshi-mittal committed:
Changed error message for non-integral datatypes

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>
